src/MonitorProfiler/ProbeInstrumentation/AssemblyProbePrep.cpp

```diff
@@ -0,0 +1,480 @@
+        probeAssemblyRefToken,
+        className.c_str(),
+        &classTypeRef));
+
+    mdMemberRef memberRef;
+    IfFailRet(pMetadataEmit->DefineMemberRef(
+        classTypeRef,
+        probeFunctionData->GetName().c_str(),
+        m_probeCache.signature.get(),
+        m_probeCache.signatureLength,
+        &memberRef));
+
+    probeMemberRef = memberRef;
+
+    return S_OK;
+}
+
+HRESULT AssemblyProbePrep::EmitNecessaryCorLibTypeTokens(
+    ModuleID moduleId,
+    COR_LIB_TYPE_TOKENS& corLibTypeTokens)
+{
+    HRESULT hr;
+
+    ComPtr<IMetaDataImport> pMetadataImport;
+    IfFailRet(m_pCorProfilerInfo->GetModuleMetaData(
+        moduleId,
+        ofRead,
+        IID_IMetaDataImport,
+        reinterpret_cast<IUnknown **>(&pMetadataImport)));
+
+    ComPtr<IMetaDataEmit> pMetadataEmit;
+    IfFailRet(m_pCorProfilerInfo->GetModuleMetaData(
+        moduleId,
+        ofRead | ofWrite,
+        IID_IMetaDataEmit,
+        reinterpret_cast<IUnknown **>(&pMetadataEmit)));
+
+    mdAssemblyRef corlibAssemblyRef;
+    IfFailRet(GetOrEmitTokenForCorLibAssemblyRef(
+        pMetadataImport,
+        pMetadataEmit,
+        corlibAssemblyRef));
+
+#define GET_OR_DEFINE_TYPE_TOKEN(name, token) \
+    IfFailRet(GetTokenForType( \
+        pMetadataImport, \
+        pMetadataEmit, \
+        corlibAssemblyRef, \
+        name, \
+        token))
+
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Boolean"), corLibTypeTokens.systemBooleanType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Byte"), corLibTypeTokens.systemByteType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Char"), corLibTypeTokens.systemCharType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Double"), corLibTypeTokens.systemDoubleType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Int16"), corLibTypeTokens.systemInt16Type);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Int32"), corLibTypeTokens.systemInt32Type);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Int64"), corLibTypeTokens.systemInt64Type);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Object"), corLibTypeTokens.systemObjectType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.SByte"), corLibTypeTokens.systemSByteType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.Single"), corLibTypeTokens.systemSingleType);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.UInt16"), corLibTypeTokens.systemUInt16Type);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.UInt32"), corLibTypeTokens.systemUInt32Type);
+    GET_OR_DEFINE_TYPE_TOKEN(_T("System.UInt64"), corLibTypeTokens.systemUInt64Type);
+
+    return S_OK;
+}
+
+HRESULT AssemblyProbePrep::GetTokenForType(
+    IMetaDataImport* pMetadataImport,
+    IMetaDataEmit* pMetadataEmit,
+    mdToken resolutionScope,
+    tstring name,
+    mdToken& typeToken)
+{
+    IfNullRet(pMetadataImport);
+    IfNullRet(pMetadataEmit);
+
+    HRESULT hr;
+
+    typeToken = mdTokenNil;
+
+    mdTypeRef typeRefToken;
+    hr = pMetadataImport->FindTypeRef(
+        resolutionScope,
+        name.c_str(),
+        &typeRefToken);
+
+    if (FAILED(hr))
```

Comment:
Do we actually need to find/Define or would Define just give back an existing token?